### PR TITLE
Streamline destructuring

### PR DESCRIPTION
**Pull request by rbuckton**
This PR cleans up some of our destructuring transform to be more consistent, along with a few additional changes:

* Converges flattening of binding pattern and assignment patterns to follow the same code paths.
* Adds to `NodeArray` a `transformFlags` property which consists of the transform flags of each node in the array **without** exclusions. This allows us to query the transform flags of each node in a node array without the need to traverse the array.
* Consolidates a few `TransformFlags` enum members to free up some space
* Added `ContainsObjectRest`/`ContainsObjectSpread` to allow us to more accurately detect esnext features.
* Changes the type check for Object rest assignment to allow any valid *lref* via a call to `checkReferenceExpression` to align with the spec proposal.
* Fixes a few cases where we unnecessarily emit the [*rval*](https://tc39.github.io/ecma262/#sec-assignment-operators-runtime-semantics-evaluation) for a `DestructuringAssignment` that goes unused.
* Cleans up `esnext.ts` to remove dependencies on `es2015` transformations.
* Moves `es2015` transformations added to `factory.ts` back into `es2015.ts`

### CODE REVIEW

**Comment by rbuckton** [comment]
// cc: @sandersn, @mhegazy

File: src/compiler/transformers/destructuring.ts [hunk 1]
```diff
@@ -3,626 +3,377 @@
 
 /*@internal*/
 namespace ts {
+    interface FlattenHost {
```
Comment (sandersn) [inline src/compiler/transformers/destructuring.ts]: probably `FlattenContext` is a better name

File: src/compiler/transformers/destructuring.ts [hunk 2]
```diff
@@ -633,27 +384,84 @@ namespace ts {
...
+
+    /** Given value: o, propName: p, pattern: { a, b, ...p } from the original statement
+     * `{ a, b, ...p } = o`, create `p = __rest(o, ["a", "b"]);`*/
+    function createRestCall(value: Expression, elements: BindingOrAssignmentElement[], computedTempVariables: Expression[], location: TextRange): Expression {
+        const propertyNames: Expression[] = [];
+        for (let i = 0; i < elements.length - 1; i++) {
+            const propertyName = getPropertyNameOfBindingOrAssignmentElement(elements[i]);
+            if (propertyName) {
+                if (isComputedPropertyName(propertyName)) {
+                    // get the temp name and put that in there instead, like `_tmp + ""`
```
Comment (sandersn) [inline src/compiler/transformers/destructuring.ts]: can you update this to just give the emit: `typeof _tmp === "symbol" ? _tmp : _tmp + ""` ? The comment I wrote is wordy and out of date to boot.

File: src/compiler/factory.ts [hunk 3]
```diff
@@ -3286,7 +3295,17 @@ namespace ts {
...
                         /*modifiers*/ undefined,
                         createVariableDeclarationList(
-                            flattenParameterDestructuring(parameter, temp, visitor, convertObjectRest)
+                            flattenDestructuringBinding(
+                                context,
+                                parameter,
+                                temp,
+                                /*skipInitializer*/ convertObjectRest,
+                                /*recordTempVariablesInLine*/ true,
+                                convertObjectRest
```
Comment (rbuckton) [inline src/compiler/factory.ts]: This is not an issue now that I've merged in the work from `isolateObjectSpread`.

File: src/compiler/factory.ts [hunk 4]
```diff
@@ -3490,12 +3506,16 @@ namespace ts {
...
-                const declarations = flattenVariableDestructuring(
+                const declarations = flattenDestructuringBinding(
+                    context,
                     firstOriginalDeclaration,
                     elementAccess,
-                    visitor,
-                    /*recordTempVariable*/ undefined,
+                    /*skipInitializer*/ false,
+                    /*recordTempVariablesInLine*/ true,
                     convertObjectRest
```
Comment (rbuckton) [inline src/compiler/factory.ts]: This is not an issue now that I've merged in the work from `isolateObjectSpread`.

File: src/compiler/transformers/destructuring.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/transformers/destructuring.ts]: Sure

File: src/compiler/transformers/destructuring.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/transformers/destructuring.ts]: Works for me.

**Pull request merged**